### PR TITLE
[F2F-297] Build /session endpoint 

### PR DESCRIPTION
**Pull request by ViveakV**
### What changed

Introduces new /session handler and TxMA SQS queue

### CODE REVIEW

File: deploy/template.yaml [hunk 1]
```diff
@@ -58,20 +58,30 @@ Mappings:
...
     build:
-      ISSUER: 'https://review-c.build.account.gov.uk'
+      ISSUER: "https://review-c.build.account.gov.uk"
+      CLIENTS: '[{"jwksEndpoint":"https://ipvstub.review-c.build.account.gov.uk/.well-known/jwks.json","clientId":"BD7B2A5D","redirectUri":"https://chspgdcmce.execute-api.eu-west-2.amazonaws.com/dev/redirect"}]'
       MESSAGERETENTIONPERIODDAYS: 345600 # Default: 4 days
+      AUTHSESSIONTTL: "604800" # 7 days in seconds
     staging:
-      ISSUER: 'https://review-c.staging.account.gov.uk'
+      ISSUER: "https://review-c.staging.account.gov.uk"
+      CLIENTS: '[{"jwksEndpoint":"https://chspgdcmce.execute-api.eu-west-2.amazonaws.com/dev/.well-known/jwks.json","clientId":"5C584572","redirectUri":"https://chspgdcmce.execute-api.eu-west-2.amazonaws.com/dev/redirect"}]'
```
Comment (richa-misra) [inline deploy/template.yaml:72]: the well known endpoints in here need to be the real IPV Core urls, please check the slack thread [here](https://gds.slack.com/archives/C0440GZHUUX/p1677760990723289?thread_ts=1677752829.023559&cid=C0440GZHUUX)

File: src/services/CicService.ts [hunk 2]
```diff
@@ -163,4 +169,79 @@ export class CicService {
     		throw new AppError("updateItem - failed: got error saving Access token details", HttpCodesEnum.SERVER_ERROR);
     	}
     }
+
+    async createAuthSession(session: ISessionItem): Promise<void> {
+    	const putSessionCommand = new PutCommand({
+    		TableName: this.tableName,
+    		Item: session,
+    	});
+    	this.logger.info("Saving session data in dynamodb" + JSON.stringify(putSessionCommand));
```
Comment (richa-misra) [inline src/services/CicService.ts:178]: json log format pls

File: src/tests/unit/data/cic-events.ts [hunk 3]
```diff
@@ -147,3 +147,53 @@ export const RESOURCE_NOT_FOUND = {
 	resource: "/invalid",
 	stageVariables: {},
 };
+
+export const VALID_SESSION = {
```
Comment (richa-misra) [inline src/tests/unit/data/cic-events.ts:151]: pls move to a separate session specific file

File: src/utils/TxmaEvent.ts [hunk 4]
```diff
@@ -32,8 +32,8 @@ export const buildCoreEventFields = (session: ISessionItem, issuer: string, sour
 	return {
 		user: {
 			user_id: session.clientId,
-			transaction_id: "",
 			persistent_session_id: session.persistentSessionId,
+			transaction_id: session.persistentSessionId,
```
Comment (richa-misra) [inline src/utils/TxmaEvent.ts:36]: let keep it null as we already got different field for persistentSessionId

File: src/utils/ValidationHelper.ts [hunk 5]
```diff
@@ -4,9 +4,10 @@ import { Logger } from "@aws-lambda-powertools/logger";
 import { HttpCodesEnum } from "./HttpCodesEnum";
 import { ISessionItem } from "../models/ISessionItem";
 import { KmsJwtAdapter } from "./KmsJwtAdapter";
-import { APIGatewayProxyEvent } from "aws-lambda";
+import { APIGatewayProxyEvent, APIGatewayProxyEventQueryStringParameters } from "aws-lambda";
```
Comment (richa-misra) [inline src/utils/ValidationHelper.ts:7]: is APIGatewayProxyEventQueryStringParameters being used?

File: src/utils/SqsClient.ts [hunk 6]
```diff
@@ -15,4 +15,4 @@ const sqsClientRaw = new SQSClient({
 
 const sqsClient = process.env.XRAY_ENABLED === "true" ? AWSXRay.captureAWSv3Client(sqsClientRaw as any) : sqsClientRaw;
 
-export { sqsClient };
+export { sqsClient, SendMessageCommand };
```
Comment (richa-misra) [inline src/utils/SqsClient.ts:18]: is SendMessageCommand being used?

File: src/tests/unit/data/cic-events.ts (hunk 3 above)
Comment (ViveakV) [reply src/tests/unit/data/cic-events.ts]: Done

File: src/services/SessionRequestProcessor.ts [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+
+	private readonly cicService: CicService;
+
+	private readonly kmsDecryptor: KmsJwtAdapter;
+
+	private readonly validationHelper: ValidationHelper;
+
+	constructor(logger: Logger, metrics: Metrics) {
+		
+		if (!process.env.SESSION_TABLE) {
```
[the remark this answers is not in the document]
Comment (ViveakV) [reply src/services/SessionRequestProcessor.ts]: Done

File: src/utils/SqsClient.ts (hunk 6 above)
Comment (ViveakV) [reply src/utils/SqsClient.ts:18]: I removed from the import in CICService and added to sqsClient in Utils, it's being used here https://github.com/alphagov/di-ipv-cri-cic-api/blob/main/src/services/CicService.ts#L11

File: src/package.json [hunk 8]
```diff
@@ -15,10 +15,12 @@
         "aws-xray-sdk-core": "^3.4.1",
         "@aws-sdk/client-sqs": "^3.276.0",
         "@aws-sdk/client-s3": "^3.120.0",
+        "aws-sdk": "^2.1150.0",
         "axios": ">=0.21.1",
         "class-validator": "0.14.0",
         "esbuild": "0.14.14",
         "node-jose": "^2.1.1",
+        "jose": "^4.11.2",
```
Comment (richa-misra) [inline src/package.json:23]: do we need both `jose` and node-jose?

File: src/package.json [hunk 9]
```diff
@@ -15,10 +15,12 @@
         "aws-xray-sdk-core": "^3.4.1",
         "@aws-sdk/client-sqs": "^3.276.0",
         "@aws-sdk/client-s3": "^3.120.0",
+        "aws-sdk": "^2.1150.0",
```
Comment (richa-misra) [inline src/package.json:18]: is this needed?

File: src/package.json (hunk 8 above)
Comment (ViveakV) [reply src/package.json]: Removed

File: src/package.json (hunk 9 above)
Comment (ViveakV) [reply src/package.json]: Removed

**Review by richa-misra — APPROVED** [review]

**Pull request merged**